apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,52 @@
+import asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import { IdSchema } from '../lib/db-types';
+import { idsEqual } from '../lib/id';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  if (!res.locals.assessment_instance) {
+    throw new Error('Assessment Instance is not pressent');
+  }
+
+  const user_session_id = await sqldb.queryOptionalRow(
+    sql.select_user_session_id,
+    { session_id: req.session.id },
+    IdSchema,
+  );
+
+  const params = {
+    ip_address: req.ip,
+    // We are passing the authn user id here. However, we are checking
+    // in the SQL query 'update_assessment_instance_fingerprint'
+    // that the authn user is the owner of the assessment. This will keep us
+    // from inadvertently recording a fingerprint change
+    // for an instructor viewing the assessment instance.
+    user_id: res.locals.authn_user.user_id,
+    user_session_id: user_session_id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
```

Comment:
If either of these headers exceed 255 characters, we'd fail to insert a client fingerprint and thus fail the entire request. We should update the `select_client_fingerprint ` and `insert_client_fingerprint` queries to explicitly cast these headers to `VARCHAR(255)` so that Postgres will correctly truncate the strings.

Ideally we'd do this in JavaScript, but Postgres limits strings by characters, whereas `.substring()` operates on bytes. So, it's better to do all truncation in Postgres for consistency.

Reply:
I have added casts for these headers in the SQL queries in commit [e6c3025](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/e6c302546f2c4cf356fafb0f205fd711c2b098ff) and [6264b21](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/6264b21af271475abb0f0bd73248f7275e216181).